data_structures/binary_tree/wavelet_tree.py

```diff
@@ -0,0 +1,214 @@
+
+
+def rank(node: Node, num: int, start: int, end: int) -> int:
+    """
+    Returns the number of occurrences of num in interval [start, end] in the list
+    """
+    if start > end:
+        return 0
+
+    rank_till_end = rank_from_start(node, num, end)  # rank of num in interval [0, end]
+    rank_before_start = rank_from_start(
+        node, num, start - 1
+    )  # rank of num in interval [0, start-1]
+
+    return rank_till_end - rank_before_start
+
+
+def quantile(node: Node, index: int, start: int, end: int) -> int:
+    """
+    Returns the index'th smallest element in interval [start, end] in the list
+    index is 0-indexed
+    """
+    if index > (end - start) or start > end:
+        return -1
+
+    # Leaf node case
+    if node.minn == node.maxx:
+        return node.minn
+
+    # Number of elements in the left subtree in interval [start, end]
+    num_elements_in_left_tree = node.map_left[end] - (
+        node.map_left[start - 1] if start else 0
+    )
+
+    if num_elements_in_left_tree > index:
+        return quantile(
+            node.left,
+            index,
+            (node.map_left[start - 1] if start else 0),
+            node.map_left[end] - 1,
+        )
+    else:
+        return quantile(
+            node.right,
+            index - num_elements_in_left_tree,
+            start - (node.map_left[start - 1] if start else 0),
+            end - node.map_left[end],
+        )
+
+
+def range_counting(
+    node: Node, start: int, end: int, start_num: int, end_num: int
+) -> int:
+    """
+    Returns the number of elememts in range [start_num, end_num]
+    in interval [start, end] in the list
+    """
+    if start > end or start_num > end_num:
+        return 0
+
+    if node.minn > end_num or node.maxx < start_num:
+        return 0
+
+    if start_num <= node.minn and node.maxx <= end_num:
+        return end - start + 1
+
+    left = range_counting(
+        node.left,
+        (node.map_left[start - 1] if start else 0),
+        node.map_left[end] - 1,
+        start_num,
+        end_num,
+    )
+    right = range_counting(
+        node.right,
+        start - (node.map_left[start - 1] if start else 0),
+        end - node.map_left[end],
+        start_num,
+        end_num,
+    )
+
+    return left + right
+
+
+def main() -> None:
+    """
+    >>> arr = [2,1,4,5,6,8,9,1,2,6,7,4,2,6,5,3,2,7]
+    >>> root = build_tree(arr)
+    >>> root
+    min_value: 1, max_value: 9
+    >>> rank(root, 6, 3, 13)
+    3
+    >>> rank(root, 2, 0, 17)
+    4
+    >>> rank(root, 9, 2 ,2)
+    0
+    >>> quantile(root, 2, 2, 5)
+    6
+    >>> quantile(root, 4, 2, 13)
+    4
+    >>> quantile(root, 0, 6, 6)
+    9
+    >>> quantile(root, 4, 2, 5)
+    -1
+    >>> range_counting(root, 1, 10, 3, 7)
+    5
+    >>> range_counting(root, 2, 2, 1, 4)
+    1
+    >>> range_counting(root, 0, 17, 1, 100)
+    18
+    >>> range_counting(root, 1, 0, 1, 100)
+    0
+    >>> range_counting(root, 0, 17, 100, 1)
+    0
+    """
```

Comment:
Please move these doctest into the methods that they test.  This will give the reader of each method an _immediate_ sense of how that method works and how it fails.  (This will also placate the bot.)